prusti-contracts-internal/src/lib.rs

```diff
@@ -52,3 +52,40 @@ pub fn extern_spec(attr: TokenStream, tokens: TokenStream) -> TokenStream {
 pub fn predicate(tokens: TokenStream) -> TokenStream {
     prusti_specs::predicate(tokens.into()).into()
 }
+/// Macro for creating models for a type.
+///
+/// Types can be annotated with the `#[model]` macro:
+/// ```ignore
+/// use std::slice::Iter;
+/// #[model]
+/// struct Iter<'a, i32> {
+///     position: usize,
+///     len: usize,
+/// }
+/// ```
+/// This creates a model for the `Iter` type defined in the standard library to be used
+/// as an abstraction in specifications. The model needs to be copyable, i.e. all fields need
+/// to be `Copy`.
+///
+/// The model can then be used in specifications:
+/// ```ignore
+/// #[ensures( result.model().position == 0 )]
+/// #[ensures( result.model().len == slice.len() )]
+/// #[trusted]
+/// fn create_iter(slice: &[i32]) -> std::slice::Iter<'_, i32> {
+///     slice.iter()
+/// }
+/// ```
+/// ## Caution
+/// The model is only to be used in specifications (pre- and postconditions of functions or methods)
+/// and never in code which will be executed. Using `.model()` will cause a panic during runtime.
+///
+/// ## Remarks
+/// * Specifications involving models can only be used in trusted functions or methods, i.e. on
+///   explicitly `#[trusted]` methods or on `#[extern_spec]`s.
+/// * Models can not be generic, i.e. creating a model for `Vec<T>` is not possible. It is however
+///   possible to model concrete generic types, i.e. `Vec<i32>` and `Vec<u32>`
```

Comment:
Good documentation, but this would probably fit better in the user guide. I'm not sure docs show up correctly on Prusti attributes in the IDE. If yes, then it should be a short explanation + a link to the user guide.

Reply:
I can at confirm that they don't show up in my IDE (clion). 
Maybe this becomes more relevant when `prusti_contracts` will be distributed crate on crates.io (#529)
For now, I will simply add this documentation to the user uid :+1:

Reply:
PR #865 created

Reply:
I incorporated your feedback in the original PR